Entities/ADoor.cpp

```diff
@@ -1,921 +1,598 @@
+
+			// TODO: Figure out how to pitch this correctly so it goes up as SpriteAnimDuration goes down.
+			/*
+			if (!m_DoorMoveSound.IsBeingPlayed()) { m_DoorMoveSound.Play(m_Pos); }
+			g_AudioMan.SetSoundPitch(&m_DoorMoveSound, LERP(0, m_InitialSpriteAnimDuration, 0, 5, m_SpriteAnimDuration));
+			*/
+
+			m_Health -= 0.4F;
+		}
+		if (m_Status == DEAD) { GibThis(Vector(), 50); }
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::UpdateSensors() {		
+		const Actor *foundActor = 0;
+		bool anySensorInput = false;
+
+		for (ADSensor &sensor : m_Sensors) {
+			foundActor = sensor.SenseActor(m_Pos, m_Rotation, m_HFlipped, m_MOID);
+			if (foundActor && foundActor->IsControllable()) {
+				anySensorInput = true;
+				// Look for enemy actors first and break the loop so we don't accidentally open the door for a friendly actor is enemies are nearby.
+				if (foundActor->GetTeam() != m_Team) {
+					CloseDoor();
+					break;
+				} else if (foundActor->GetTeam() == m_Team) {					
+					OpenDoor();
+				}
+			}
+		}
+		// If no sensor input, return to default state if it's time. This won't override any actions that the sensors started
+		if (!anySensorInput && m_ResetDefaultTimer.IsPastSimMS(m_ResetDefaultDelay)) {
+			if (m_ClosedByDefault && m_DoorState == OPEN) {
+				CloseDoor();
+			} else if (!m_ClosedByDefault && m_DoorState == CLOSED) {
+				OpenDoor();
+			}
+		}
+		m_SensorTimer.Reset();
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ADoor::UpdateDoorAttachable() {
+		m_Door->SetHFlipped(m_HFlipped);
+		switch (m_DoorState) {
+			case OPEN:
+				m_Door->SetJointPos(m_Pos + m_OpenOffset.GetXFlipped(m_HFlipped) * m_Rotation);
+				m_Door->SetRotAngle(m_Rotation.GetRadAngle() + (m_OpenAngle * GetFlipFactor()));
+				break;
+			case OPENING:
+				if (!m_DoorMoveSound.IsBeingPlayed()) { m_DoorMoveSound.Play(m_Pos); }
+
+				if (m_DoorMoveTimer.IsPastSimMS(m_DoorMoveTime)) {
+					m_ResetDefaultTimer.Reset();
+
+					m_Door->SetJointPos(m_Pos + m_OpenOffset.GetXFlipped(m_HFlipped) * m_Rotation);
+					m_Door->SetRotAngle(m_Rotation.GetRadAngle() + (m_OpenAngle * GetFlipFactor()));
+
+					// Draw the material representation of the door to the terrain in the open state
+					if (m_DrawMaterialLayerWhenOpen) { DrawDoorMaterial(); }
+
+					m_DoorMoveSound.Stop();
+					m_DoorMoveEndSound.Play(m_Pos);
+
+					m_DoorState = OPEN;
+				} else {
+					// Find interpolated positions and angles for opening
+					Vector openingOffset(LERP(0, m_DoorMoveTime, m_ClosedOffset.m_X, m_OpenOffset.m_X, m_DoorMoveTimer.GetElapsedSimTimeMS()), LERP(0, m_DoorMoveTime, m_ClosedOffset.m_Y, m_OpenOffset.m_Y, m_DoorMoveTimer.GetElapsedSimTimeMS()));
+					// TODO: Make this work across rotation 0
```

Comment:
Any clue what this TODO is about, if it's still relevant or how to complete it?

Reply:
Might have to do with doors that rotate between positive and negative angles, like from 90 to -90 or something like that.
Don't know if we really need that.

Reply:
Seems like the easy solve would be, when reading in ini or whatever, if an angle is < 0, put a while loop at adds 2pi to it until it's a positive angle. (e.g. -90 deg + 2*180 deg = 270 deg). That way your angles are always >= 0 and you don't have to worry about it.

Reply:
Fix is more complicated than I thought, since just converting +- angles to 0-2pi ones would not necessarily respect rotation direction. I'm leaving this one be but doing a reader abort if angle < 0 so modders know they screwed something up.